catroid/src/main/java/org/catrobat/catroid/formulaeditor/SensorLoudness.kt

```diff
@@ -0,0 +1,113 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import org.catrobat.catroid.soundrecorder.SoundRecorder
+import java.io.IOException
+import java.util.ArrayList
+
+class SensorLoudness(private var recorder: SoundRecorder = SoundRecorder()) {
+    private val listenerList: MutableList<SensorCustomEventListener> = ArrayList()
+    private var lastValue = 0.0
+    private var coroutine: Job? = null
+
+    @JvmOverloads
+    fun runLoudnessSensorAsync(
+        scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
+        dispatcherOnFinished: CoroutineDispatcher = Dispatchers.Main
+    ) {
+        coroutine = scope.launch {
+            while (isActive) {
+                val loudness = SCALE_RANGE / MAX_AMP_VALUE * recorder.maxAmplitude
+                if (loudness != lastValue && loudness != 0.0) {
+                    lastValue = loudness
+                    val event = SensorCustomEvent(Sensors.LOUDNESS, loudness)
+                    withContext(dispatcherOnFinished) {
+                        for (listener in listenerList) {
+                            listener.onCustomSensorChanged(event)
+                        }
+                    }
+                    delay(UPDATE_INTERVAL.toLong())
+                }
+            }
+        }
+    }
+
+    @Synchronized
+    fun registerListener(listener: SensorCustomEventListener): Boolean {
+        if (listenerList.contains(listener)) return true
+        listenerList.add(listener)
+        if (!recorder.isRecording) {
+            try {
+                recorder.start()
+                runLoudnessSensorAsync()
+            } catch (ioException: IOException) {
+                Log.d(TAG, "Could not start recorder", ioException)
+                listenerList.remove(listener)
+                recorder = SoundRecorder()
+                return false
+            } catch (e: IllegalStateException) {
+                Log.d(TAG, "Could not start recorder", e)
+                listenerList.remove(listener)
+                recorder = SoundRecorder()
+                return false
+            }
+        }
+        return true
+    }
+
+    @Synchronized
+    fun unregisterListener(listener: SensorCustomEventListener?) {
+        if (!listenerList.contains(listener)) return
+        listenerList.remove(listener)
+        coroutine?.cancel()
+        if (listenerList.size == 0) {
+            if (recorder.isRecording) {
+                try {
+                    recorder.stop()
+                } catch (ioException: IOException) {
+                    // ignored, nothing we can do
+                    Log.d(TAG, "Could not stop recorder", ioException)
+                }
+                recorder = SoundRecorder()
+            }
+            lastValue = 0.0
+        }
+    }
+
+    companion object {
+        private const val UPDATE_INTERVAL = 50
+        private const val SCALE_RANGE = 100.0
+        private const val MAX_AMP_VALUE = 32_767.0
```

Comment:
Refactor to private compile-time constants, please.
https://kotlinlang.org/docs/properties.html#compile-time-constants